language/move-prover/tests/sources/invariants.move

```diff
@@ -0,0 +1,112 @@
+module Invariants {
+
+    // General invariant checking.
+
+    struct R {
+        x: u64
+    }
+
+    spec struct R {
+      // We must always have a value greater one.
+      invariant x > 1;
+
+      // When we update via a reference, the new value must be smaller or equal the old one.
+      invariant update x <= old(x);
```

Comment:
Same question as above but now for invariants. What does old(x) mean?

Reply:
What can the invariant expression refer to in general?

Reply:
> All fields and global spec vars.

At the moment, global spec vars are not tied to any particular resource type.  As far as I understand, they were introduced simply as a mechanism to introduce auxiliary state and there is no notion of "soundness" associated with them.  However, my understanding is that there is a clear notion of "soundness" associated with resource invariants.  My worry is that if a resource invariant is allowed to refer to any global spec variable, we will compromise the soundness.  I think resource invariant on T can be made to be sound even if it refers to a global variable x provided x is associated with type T and is not updated in a different type U.